File: src/lib/boardingPass.ts
/**
 * Slice-A — signed boarding-pass QR.
 *
 * The QR encodes a JSON payload describing one trip leg, plus an HMAC-SHA256
 * signature over a stable canonical form of that payload. A verifier (e.g.
 * `KioskSimulator`) recomputes the HMAC with the same shared secret and
 * compares constant-time. Tampered or replayed payloads (after expiry)
 * fail verification.
 *
 * This is real cryptography (Web Crypto subtle.HMAC, SHA-256) — distinct
 * from the toy `simSign` in `tokenService.ts` which is left unchanged for
 * the existing identity-verification flow's contract.
 *
 * Caveat: this is NOT a real airline-issued IATA BCBP. Real boarding
 * passes are signed by the operating carrier's keypair. The payload still
 * carries `kind: "globeid.bp.v1"` + `isDemoData: true` so any genuine
 * gate verifier will reject it. Within the GlobeID app and the bundled
 * `KioskSimulator` mock, signature verification is real and meaningful.
 */
const SECRET: string =
  (typeof process !== "undefined" && process.env?.GLOBE_BP_SECRET) ||
  (typeof import.meta !== "undefined" && (import.meta as { env?: { VITE_BOARDING_PASS_SECRET?: string } }).env?.VITE_BOARDING_PASS_SECRET) ||
  "globe-dev-bp-secret-change-me";

const KIND = "globeid.bp.v1";

export interface BoardingPassPayload {
  kind: typeof KIND;
  isDemoData: true;
  passenger: string;
  passportLast4: string | null;
  flightNumber: string;
  airline: string;
  fromIata: string;
  toIata: string;
  scheduledDate: string;
  legId: string;
  tripId: string | null;
  iat: number; // issued-at unix ms
  exp: number; // expiry unix ms (departure +24h)
  appOrigin: string;
}

export interface SignedBoardingPass {
  payload: BoardingPassPayload;
  qrText: string;
}

export interface VerificationOk {
  valid: true;
  payload: BoardingPassPayload;
}

export interface VerificationFailed {
  valid: false;
  payload: BoardingPassPayload | null;
  error: string;
}

export type VerificationResult = VerificationOk | VerificationFailed;

export interface IssueArgs {
  passenger: string;
  passportNo: string | null;
  flightNumber: string;
  airline: string;
  fromIata: string;
  toIata: string;
  scheduledDate: string;
  legId: string;
  tripId: string | null;
  /** Override expiry (unix ms). Defaults to scheduledDate + 24h. */
  exp?: number;
  /** Override iat (unix ms). Defaults to Date.now(). */
  iat?: number;
}

function getCrypto(): Crypto {
  if (typeof globalThis !== "undefined" && globalThis.crypto?.subtle) {
    return globalThis.crypto;
  }
  throw new Error("Web Crypto subtle is unavailable in this environment");
}

function utf8Encode(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

function bytesToBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function base64UrlToBytes(b64: string): Uint8Array {
  const padded = b64.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - (b64.length % 4)) % 4);
  const binary = atob(padded);
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

/** Stable JSON form: keys sorted, no whitespace. */
function canonicalize(payload: BoardingPassPayload): string {
  const ordered: Record<string, unknown> = {};
  for (const k of (Object.keys(payload) as Array<keyof BoardingPassPayload>).sort()) {
    ordered[k as string] = payload[k];
  }
  return JSON.stringify(ordered);
}

async function importHmacKey(secret: string): Promise<CryptoKey> {
  return getCrypto().subtle.importKey(
    "raw",
    utf8Encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign", "verify"],
  );
}

async function hmacSign(secret: string, data: string): Promise<string> {
  const key = await importHmacKey(secret);
  const sigBuf = await getCrypto().subtle.sign("HMAC", key, utf8Encode(data));
  return bytesToBase64Url(new Uint8Array(sigBuf));
}

async function hmacVerify(secret: string, data: string, sigB64: string): Promise<boolean> {
  const key = await importHmacKey(secret);
  let sig: Uint8Array;
  try {
    sig = base64UrlToBytes(sigB64);
  } catch {
    return false;
  }
  return getCrypto().subtle.verify("HMAC", key, sig, utf8Encode(data));
}

function deriveExp(scheduledDate: string): number {
  // Departure date at UTC noon + 24h. Wide enough for layover lounges
  // and red-eye departures that span midnight UTC.
  const t = Date.parse(`${scheduledDate}T12:00:00Z`);
  if (Number.isNaN(t)) return Date.now() + 7 * 24 * 3_600_000;
  return t + 24 * 3_600_000;
}

export async function issueBoardingPass(
  args: IssueArgs,
  secretOverride?: string,
): Promise<SignedBoardingPass> {
  const passportLast4 =
    args.passportNo && args.passportNo.length >= 4 ? args.passportNo.slice(-4) : null;
  const now = args.iat ?? Date.now();
  const payload: BoardingPassPayload = {
    kind: KIND,
    isDemoData: true,
    passenger: args.passenger,
    passportLast4,
    flightNumber: args.flightNumber,
    airline: args.airline,
    fromIata: args.fromIata,
    toIata: args.toIata,
    scheduledDate: args.scheduledDate,
    legId: args.legId,
    tripId: args.tripId,
    iat: now,
    exp: args.exp ?? deriveExp(args.scheduledDate),
    appOrigin: typeof window !== "undefined" ? window.location.origin : "globeid",
  };
  const canonical = canonicalize(payload);
  const sig = await hmacSign(secretOverride ?? SECRET, canonical);
  const envelope = { p: payload, s: sig };
  return { payload, qrText: JSON.stringify(envelope) };
}

export async function verifyBoardingPass(
  qrText: string,
  secretOverride?: string,
): Promise<VerificationResult> {
  let envelope: { p?: unknown; s?: unknown };
  try {
    envelope = JSON.parse(qrText);
  } catch {
    return { valid: false, payload: null, error: "Malformed QR payload" };
  }
  if (!envelope || typeof envelope !== "object" || typeof envelope.s !== "string") {
    return { valid: false, payload: null, error: "Missing signature" };
  }
  const candidate = envelope.p as Partial<BoardingPassPayload>;
  if (!candidate || candidate.kind !== KIND) {
    return { valid: false, payload: null, error: "Not a GlobeID boarding pass" };
  }
  const required: Array<keyof BoardingPassPayload> = [
    "kind",
    "isDemoData",
    "passenger",
    "flightNumber",
    "airline",
    "fromIata",
    "toIata",
    "scheduledDate",
    "legId",
    "iat",
    "exp",
    "appOrigin",
  ];
  for (const k of required) {
    if (candidate[k] === undefined) {
      return { valid: false, payload: null, error: `Missing field: ${k}` };
    }
  }
  const payload = candidate as BoardingPassPayload;
  const canonical = canonicalize(payload);
  const ok = await hmacVerify(secretOverride ?? SECRET, canonical, envelope.s);
  if (!ok) return { valid: false, payload, error: "Signature mismatch" };
  if (Date.now() > payload.exp) return { valid: false, payload, error: "Pass expired" };
  return { valid: true, payload };
}

File: flutter_app/lib/domain/boarding_pass.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Dart port of `src/lib/boardingPass.ts`. Real HMAC-SHA256 sign +
/// constant-time verify with a base64url signature envelope.
const _kind = 'globeid.bp.v1';
const _defaultSecret = String.fromEnvironment(
  'BOARDING_PASS_SECRET',
  defaultValue: 'globe-dev-bp-secret-change-me',
);

class BoardingPassPayload {
  BoardingPassPayload({
    required this.passenger,
    required this.passportLast4,
    required this.flightNumber,
    required this.airline,
    required this.fromIata,
    required this.toIata,
    required this.scheduledDate,
    required this.legId,
    required this.tripId,
    required this.iat,
    required this.exp,
    required this.appOrigin,
  });

  final String passenger;
  final String? passportLast4;
  final String flightNumber;
  final String airline;
  final String fromIata;
  final String toIata;
  final String scheduledDate;
  final String legId;
  final String? tripId;
  final int iat;
  final int exp;
  final String appOrigin;

  Map<String, dynamic> toJson() => {
        'kind': _kind,
        'isDemoData': true,
        'passenger': passenger,
        'passportLast4': passportLast4,
        'flightNumber': flightNumber,
        'airline': airline,
        'fromIata': fromIata,
        'toIata': toIata,
        'scheduledDate': scheduledDate,
        'legId': legId,
        'tripId': tripId,
        'iat': iat,
        'exp': exp,
        'appOrigin': appOrigin,
      };

  factory BoardingPassPayload.fromJson(Map<String, dynamic> j) =>
      BoardingPassPayload(
        passenger: j['passenger'] as String,
        passportLast4: j['passportLast4'] as String?,
        flightNumber: (j['flightNumber'] as String?) ?? 'TBD',
        airline: j['airline'] as String,
        fromIata: j['fromIata'] as String,
        toIata: j['toIata'] as String,
        scheduledDate: j['scheduledDate'] as String,
        legId: j['legId'] as String,
        tripId: j['tripId'] as String?,
        iat: (j['iat'] as num).toInt(),
        exp: (j['exp'] as num).toInt(),
        appOrigin: j['appOrigin'] as String,
      );
}

class SignedBoardingPass {
  SignedBoardingPass({required this.payload, required this.qrText});
  final BoardingPassPayload payload;
  final String qrText;
}

class BoardingPassVerification {
  BoardingPassVerification.ok(this.payload)
      : valid = true,
        error = null;
  BoardingPassVerification.fail(this.error, [this.payload]) : valid = false;
  final bool valid;
  final BoardingPassPayload? payload;
  final String? error;
}

String _canonicalize(Map<String, dynamic> payload) {
  final keys = payload.keys.toList()..sort();
  final ordered = <String, dynamic>{for (final k in keys) k: payload[k]};
  return jsonEncode(ordered);
}

String _base64Url(List<int> bytes) =>
    base64Url.encode(bytes).replaceAll('=', '');

int _deriveExp(String scheduledDate) {
  try {
    final t =
        DateTime.parse('${scheduledDate}T12:00:00Z').millisecondsSinceEpoch;
    return t + 24 * 3600 * 1000;
  } catch (_) {
    return DateTime.now().millisecondsSinceEpoch + 7 * 24 * 3600 * 1000;
  }
}

SignedBoardingPass issueBoardingPass({
  required String passenger,
  String? passportNo,
  String? flightNumber,
  required String airline,
  required String fromIata,
  required String toIata,
  required String scheduledDate,
  required String legId,
  String? tripId,
  int? iat,
  int? exp,
  String? secretOverride,
  String appOrigin = 'globeid',
}) {
  final last4 = (passportNo != null && passportNo.length >= 4)
      ? passportNo.substring(passportNo.length - 4)
      : null;
  final payload = BoardingPassPayload(
    passenger: passenger,
    passportLast4: last4,
    flightNumber: flightNumber ?? 'TBD',
    airline: airline,
    fromIata: fromIata,
    toIata: toIata,
    scheduledDate: scheduledDate,
    legId: legId,
    tripId: tripId,
    iat: iat ?? DateTime.now().millisecondsSinceEpoch,
    exp: exp ?? _deriveExp(scheduledDate),
    appOrigin: appOrigin,
  );
  final canonical = _canonicalize(payload.toJson());
  final secret = secretOverride ?? _defaultSecret;
  final sig = Hmac(sha256, utf8.encode(secret)).convert(utf8.encode(canonical));
  final envelope = {'p': payload.toJson(), 's': _base64Url(sig.bytes)};
  return SignedBoardingPass(payload: payload, qrText: jsonEncode(envelope));
}

BoardingPassVerification verifyBoardingPass(String qrText,
    {String? secretOverride}) {
  Map<String, dynamic> envelope;
  try {
    final decoded = jsonDecode(qrText);
    if (decoded is! Map<String, dynamic>) {
      return BoardingPassVerification.fail('Malformed QR payload');
    }
    envelope = decoded;
  } catch (_) {
    return BoardingPassVerification.fail('Malformed QR payload');
  }
  final sig = envelope['s'];
  final candidate = envelope['p'];
  if (sig is! String || candidate is! Map<String, dynamic>) {
    return BoardingPassVerification.fail('Missing signature');
  }
  if (candidate['kind'] != _kind) {
    return BoardingPassVerification.fail('Not a GlobeID boarding pass');
  }
  for (final k in const [
    'kind',
    'isDemoData',
    'passenger',
    'flightNumber',
    'airline',
    'fromIata',
    'toIata',
    'scheduledDate',
    'legId',
    'iat',
    'exp',
    'appOrigin',
  ]) {
    if (!candidate.containsKey(k)) {
      return BoardingPassVerification.fail('Missing field: $k');
    }
  }
  final payload = BoardingPassPayload.fromJson(candidate);
  final canonical = _canonicalize(payload.toJson());
  final secret = secretOverride ?? _defaultSecret;
  final expected =
      Hmac(sha256, utf8.encode(secret)).convert(utf8.encode(canonical));
  final expectedSig = _base64Url(expected.bytes);
  // Constant-time string comparison.
  if (expectedSig.length != sig.length) {
    return BoardingPassVerification.fail('Signature mismatch', payload);
  }
  var diff = 0;
  for (var i = 0; i < expectedSig.length; i++) {
    diff |= expectedSig.codeUnitAt(i) ^ sig.codeUnitAt(i);
  }
  if (diff != 0) {
    return BoardingPassVerification.fail('Signature mismatch', payload);
  }
  if (DateTime.now().millisecondsSinceEpoch > payload.exp) {
    return BoardingPassVerification.fail('Pass expired', payload);
  }
  return BoardingPassVerification.ok(payload);
}
